src/ripple/app/misc/NetworkOPs.cpp

```diff
@@ -979,43 +997,71 @@ void NetworkOPsImp::apply (std::unique_lock<std::mutex>& batchLock)
     batchLock.unlock();
 
     {
-        Ledger::pointer ledger;
+        std::shared_ptr<Ledger> newOL;
         bool applied = false;
-        boost::optional<MetaView> accum;
+        boost::optional<OpenView> accum;
         auto lock = beast::make_lock(getApp().getMasterMutex());
         {
             std::lock_guard <std::recursive_mutex> lock (
                 m_ledgerMaster.peekMutex());
-            ledger = m_ledgerMaster.getCurrentLedgerHolder().getMutable();
-            accum.emplace(*ledger, tapNONE);
+        #if RIPPLE_OPEN_LEDGER
+            auto const oldOL = m_ledgerMaster.getCurrentLedgerHolder().get();
+            getApp().openLedger().verify(*oldOL, "apply before");
+        #endif
+            newOL = m_ledgerMaster.getCurrentLedgerHolder().getMutable();
+            accum.emplace(&*newOL);
             for (TransactionStatus& e : transactions)
             {
-                ViewFlags flags = tapNONE;
+                ApplyFlags flags = tapNONE;
                 flags = flags | tapNO_CHECK_SIGN;
                 if (e.admin)
                     flags = flags | tapADMIN;
+
+            #if RIPPLE_OPEN_LEDGER
+                auto const sle1 = accum->read(
+                    keylet::account(e.transaction->getSTransaction()->getAccountID(sfAccount)));
+            #endif
+
                 std::tie (e.result, e.applied) =
                     ripple::apply (*accum,
                         *e.transaction->getSTransaction(), flags,
                             getConfig(), deprecatedLogs().journal(
                                 "NetworkOPs"));
                 applied |= e.applied;
+
+            #if RIPPLE_OPEN_LEDGER
+                auto const sle2 = getApp().openLedger().current()->read(keylet::account(e.transaction->getSTransaction()->getAccountID(sfAccount)));
+                // VFALCO Should do the loop inside modify()
+                getApp().openLedger().modify(
+                    [&](OpenView& view, beast::Journal j)
+                    {
```

Comment:
The basic idea is that there's some stuff you have to do before and after you modify an open ledger. To ensure the invariants are respected, there's a `modify` function that does the before stuff, does the modifications you ask it to do, and then does the after stuff. The alternative would be to have a modification object whose destructor completed the modification.